tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerOperationBuilder.java

```diff
@@ -277,22 +257,18 @@ private static void requireTensorHandle(TFE_TensorHandle handle) {
     }
   }
 
-  private static TFE_Op allocate(TFE_Context ctxHandle, String type) {
+  private static TFE_Op allocate(TFE_Context ctxHandle, String type, PointerScope outputScope) {
```

Comment:
Can you please elaborate more on how `PointerScope` is working? 

For instance, the scope define in this method is not passed anywhere so I assume that the pointer created by `TFE_Op.newOp` is automatically attached to it on creation? Then, since we are already in the scope of another `PointerScope` (the one passed as `outputScope`), would it be automatically attached to it as well if we were not declaring a new one here (and shouldn't that be enough)? 

If it is not required, I think it would make the code easier to follow not having so many scopes declared during the process.

Reply:
In other words, instead of :
```java
void create() {
  try (PointerScope scope = new PointerScope()) {
    allocate(scope);
  }
}

private void allocate(PointerScope outputScope) {
  try (PointerScope scope = new PointerScope()) {
    Pointer p = new Pointer();
    outputScope.attach(p);
  }
}
```
, can't we simply do this?
```java
void create() {
  try (PointerScope scope = new PointerScope()) {
    allocate(scope);
  }
}

private void allocate(PointerScope outputScope) {
  return new Pointer();
}
```

Reply:
> Can you please elaborate more on how `PointerScope` is working?

- http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/
- https://github.com/karllessard/tensorflow-java-sandbox/issues/3

> For instance, the scope define in this method is not passed anywhere so I assume that the pointer created by `TFE_Op.newOp` is automatically attached to it on creation? Then, since we are already in the scope of another `PointerScope` (the one passed as `outputScope`), would it be automatically attached to it as well if we were not declaring a new one here (and shouldn't that be enough)?

Yes, that's the general idea. We need the new scopes to clean up any local resources that we don't want lying around.

> If it is not required, I think it would make the code easier to follow not having so many scopes declared during the process.

Surely, but I don't see an easier way of doing this, short of doing reference counting manually. Would you like to go that way? :) FWIW, to make it prettier, this pattern should really be integrated into the Java language. There's still most likely some more that can be done at the library level. I'll be updating JavaCPP with any ideas that help, for sure, but I'm not having any more for now...

Reply:
> can't we simply do this?

Yes, but this is not what we have. We've got at least the `status` object to deallocate here.

Reply:
I'm refactoring this a bit to get rid of those temporary scopes by using the scopes already available in the session objects. I'll push the changes soon.

Reply:
But the status will be deallocated ultimately by the outer scope as well, right?

I suggest then that we opt for simplicity and that we let the outer scopes release resources allocated by their private helper methods, which will probably end up being inlined anyway if we can them simple enough.

Reply:
Sure, we can refactor some more and start inlining those private methods. We don't need to keep them separate now that we're not using writing JNI anymore. Anyway, let me know if it's good enough with the latest changes.

Reply:
Yep, looks good!